chain-evm/src/precompiles/blake2.rs

```diff
@@ -0,0 +1,280 @@
+use evm::{Context, ExitError};
+
+use super::prelude::{mem, Address, Borrowed, TryInto};
+use super::{EvmPrecompileResult, Precompile, PrecompileOutput};
+
+/// Blake2 costs.
+mod costs {
+    /// Cost per round of Blake2 F.
+    pub(super) const F_ROUND: u64 = 1;
+}
+
+/// Blake2 constants.
+mod consts {
+    pub(super) const INPUT_LENGTH: usize = 213;
+}
+
+pub(super) struct Blake2F;
+
+impl Blake2F {
+    pub(super) const ADDRESS: Address = super::make_address(0, 9);
+}
+
+impl Precompile for Blake2F {
+    fn required_gas(input: &[u8]) -> Result<u64, ExitError> {
+        let (int_bytes, _) = input.split_at(mem::size_of::<u32>());
+        Ok(u64::from(u32::from_be_bytes(
+            int_bytes.try_into().expect("cannot fail"),
+        )) * costs::F_ROUND)
+    }
+
+    /// The compression function of the blake2 algorithm.
+    ///
+    /// Takes as an argument the state vector `h`, message block vector `m` (the last block is padded
+    /// with zeros to full block size, if required), 2w-bit offset counter `t`, and final block
+    /// indicator flag `f`. Local vector v[0..15] is used in processing. F returns a new state vector.
+    /// The number of rounds, `r`, is 12 for BLAKE2b and 10 for BLAKE2s. Rounds are numbered from 0 to
+    /// r - 1.
+    ///
+    /// See: https://eips.ethereum.org/EIPS/eip-152
+    /// See: https://etherscan.io/address/0000000000000000000000000000000000000009
+    fn run(
+        input: &[u8],
+        target_gas: Option<u64>,
+        _context: &Context,
+        _is_static: bool,
+    ) -> EvmPrecompileResult {
+        if input.len() != consts::INPUT_LENGTH {
+            return Err(ExitError::Other(Borrowed("ERR_BLAKE2F_INVALID_LEN")));
+        }
+
+        let cost = Self::required_gas(input)?;
+        if let Some(target_gas) = target_gas {
+            if cost > target_gas {
+                return Err(ExitError::OutOfGas);
+            }
+        }
+
+        let mut rounds_bytes = [0u8; 4];
```

Comment:
maybe better use our `Codec` for all these reading operations ?
or at least move it into another separate functions ?

Reply:
Mmm, not sure that applies here. This is the format EVM code expects to interact with a precompile function

Reply:
Good calls, thanks. This code was brought in as-is from the Aurora Engine (`CC0` licensed), with the intention to refactor it later. I think that these comments should be moved into a new issue @eugene-babichenko .